pkg/models/devops/s2ibinary_handler.go

```diff
@@ -0,0 +1,198 @@
+package devops
+
+import (
+	"code.cloudfoundry.org/bytefmt"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/emicklei/go-restful"
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"kubesphere.io/kubesphere/pkg/apis/devops/v1alpha1"
+	"kubesphere.io/kubesphere/pkg/informers"
+	"kubesphere.io/kubesphere/pkg/simple/client/k8s"
+	"kubesphere.io/kubesphere/pkg/simple/client/s2is3"
+	"mime/multipart"
+	"net/http"
+	"reflect"
+	"time"
+)
+
+const (
+	GetS2iBinaryURL = "http://ks-apiserver.kubesphere-system.svc/kapis/devops.kubesphere.io/v1alpha2/namespaces/%s/s2ibinaries/%s/file/%s"
+)
+
+func UploadS2iBinary(namespace, name, md5 string, fileHeader *multipart.FileHeader) (*v1alpha1.S2iBinary, error) {
+	binFile, err := fileHeader.Open()
+	if err != nil {
+		glog.Errorf("%+v", err)
+		return nil, err
+	}
+	defer binFile.Close()
+
+	origin, err := informers.KsSharedInformerFactory().Devops().V1alpha1().S2iBinaries().Lister().S2iBinaries(namespace).Get(name)
+	if err != nil {
+		glog.Errorf("%+v", err)
+		return nil, err
+	}
+	//Check file is uploading
+	if origin.Status.Phase == v1alpha1.StatusUploading {
+		err := restful.NewError(http.StatusConflict, "file is uploading, please try later")
+		glog.Error(err)
+		return nil, err
+	}
+	copy := origin.DeepCopy()
+	copy.Spec.MD5 = md5
+	copy.Spec.Size = bytefmt.ByteSize(uint64(fileHeader.Size))
+	copy.Spec.FileName = fileHeader.Filename
+	copy.Spec.DownloadURL = fmt.Sprintf(GetS2iBinaryURL, namespace, name, copy.Spec.FileName)
+	if origin.Status.Phase == v1alpha1.StatusReady && reflect.DeepEqual(origin, copy) {
+		return origin, nil
+	}
+
+	//Set status Uploading to lock resource
+	origin, err = SetS2iBinaryStatus(origin, v1alpha1.StatusUploading)
+	if err != nil {
+		err := restful.NewError(http.StatusConflict, fmt.Sprintf("could not set status: %+v", err))
+		glog.Error(err)
+		return nil, err
+	}
+	copy = origin.DeepCopy()
+	copy.Spec.MD5 = md5
+	copy.Spec.Size = bytefmt.ByteSize(uint64(fileHeader.Size))
+	copy.Spec.FileName = fileHeader.Filename
+	copy.Spec.DownloadURL = fmt.Sprintf(GetS2iBinaryURL, namespace, name, copy.Spec.FileName)
+	s3session := s2is3.Session()
+	if s3session == nil {
```

Comment:
if unable to connect to storage service, already set object status to StatusUploading, return error when object in StatusUplaoding, so the object will remain in StatusUploading forever.